bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/RuntimeRegistry.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.ballerina.runtime.internal.scheduling;
+
+import io.ballerina.runtime.api.PredefinedTypes;
+import io.ballerina.runtime.api.async.StrandMetadata;
+import io.ballerina.runtime.api.creators.ErrorCreator;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.values.BError;
+import io.ballerina.runtime.api.values.BFunctionPointer;
+import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.runtime.internal.values.FutureValue;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.Stack;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * The registry for runtime dynamic listeners and stop handlers.
+ *
+ * @since 2201.2.0
+ */
+public class RuntimeRegistry {
+
+    private final Scheduler scheduler;
+    private final Set<BObject> listenerSet = new HashSet<>();
+    private final Stack<BFunctionPointer<?, ?>> stopHandlerStack = new Stack<>();
+
+    public RuntimeRegistry(Scheduler scheduler) {
+        this.scheduler = scheduler;
+    }
+
+    public synchronized void registerListener(BObject listener) {
+        listenerSet.add(listener);
+        scheduler.setImmortal(true);
+    }
+
+    public synchronized void deregisterListener(BObject listener) {
+        listenerSet.remove(listener);
+        if (!scheduler.isListenerDeclarationFound() && listenerSet.isEmpty()) {
+            scheduler.setImmortal(false);
+        }
+    }
+
+    public synchronized void stopListeners(Strand strand) {
```

Comment:
Lets add stopdynamic and merge both listener graceful stop and stop handler graceful stop